actors/evm/src/interpreter/precompiles.rs

```diff
@@ -71,24 +71,36 @@ fn read_u256_infalliable(buf: &[u8], start: usize) -> U256 {
 }
 
 fn ec_recover<RT: Primitives>(rt: &RT, input: &[u8]) -> PrecompileResult {
-    let mut buf = [0u8; 128];
-    buf[..input.len().min(128)].copy_from_slice(&input[..input.len().min(128)]);
-
+    if input.len() != 128 {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
     let mut hash = [0u8; SECP_SIG_MESSAGE_HASH_SIZE];
     let mut sig = [0u8; SECP_SIG_LEN];
 
-    hash.copy_from_slice(&input[..32]);
-    sig.copy_from_slice(&input[64..]); // TODO this assumes input is exactly 65 bytes which would panic if incorrect
+    hash.copy_from_slice(&input[0..32]);
+    sig[..64].copy_from_slice(&input[64..128]);
+    let recovery_byte = input[63];
 
-    // recovery byte means a single byte value is 32 bytes long, sad
-    if input[32..63] != [0u8; 31] || !matches!(input[63], 23 | 28) {
+    // recovery byte is a single byte value but is represented with 32 bytes, sad
+    let p = if input[32..63] == [0u8; 31] && matches!(recovery_byte, 27 | 28) {
+        recovery_byte - 27
+    } else {
         return Ok(Vec::new());
-    }
-    sig[64] = input[63] - 27;
+    };
+
+    sig[64] = p;
+
+    let pubkey = if let Ok(key) = rt.recover_secp_public_key(&hash, &sig) {
```

Comment:
We should double-check with a key that fails https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/crypto/crypto.go#L272